File: asr/harmony/sdk/src/main/ets/com/amphion/asr/SpeakerPcmBuffers.ts
export class SpeakerPcmBuffers {
  private maxSamples: number;
  private speakerVadParts: Float32Array[] = [];
  private fallbackParts: Float32Array[] = [];
  private speakerVadSampleCount: number = 0;
  private fallbackSampleCount: number = 0;

  constructor(maxSamples: number) {
    this.maxSamples = Math.max(1, Math.round(maxSamples));
  }

  observe(samples: Float32Array, captureSpeakerVad: boolean, captureFallback: boolean): void {
    if (captureSpeakerVad) {
      const retained = Math.min(samples.length, this.maxSamples - this.speakerVadSampleCount);
      if (retained > 0) {
        this.speakerVadParts.push(samples.slice(0, retained));
        this.speakerVadSampleCount += retained;
      }
    }
    if (captureFallback) {
      const retained = Math.min(samples.length, this.maxSamples - this.fallbackSampleCount);
      if (retained > 0) {
        this.fallbackParts.push(samples.slice(0, retained));
        this.fallbackSampleCount += retained;
      }
    }
  }

  speakerVadLength(): number {
    return this.speakerVadSampleCount;
  }

  speakerVadTail(n: number): Float32Array {
    const all = concatFloat32(this.speakerVadParts, this.speakerVadSampleCount);
    if (all.length <= n) return all;
    return all.slice(all.length - n, all.length);
  }

  fallbackSamples(): Float32Array {
    return concatFloat32(this.fallbackParts, this.fallbackSampleCount);
  }

  clearNativeSegment(): void {
    this.speakerVadParts = [];
    this.speakerVadSampleCount = 0;
  }

  clearPublicUtterance(): void {
    this.fallbackParts = [];
    this.fallbackSampleCount = 0;
  }

  clearAll(): void {
    this.clearNativeSegment();
    this.clearPublicUtterance();
  }
}

function concatFloat32(parts: Float32Array[], total: number): Float32Array {
  const out = new Float32Array(total);
  let offset = 0;
  for (let i = 0; i < parts.length; i++) {
    out.set(parts[i], offset);
    offset += parts[i].length;
  }
  return out;
}

File: asr/android/sdk/src/main/java/com/amphion/asr/internal/SpeakerPcmBuffers.kt
package com.amphion.asr.internal

/**
 * Keeps native-stream Speaker VAD audio separate from public-utterance fallback audio.
 *
 * A token-only native endpoint clears the first buffer after the stream reset, but the second buffer
 * must survive until a public final so two short native segments can still reach the scoring minimum.
 */
internal class SpeakerPcmBuffers(maxSamples: Int) {
    private val maximumSamples = maxSamples.coerceAtLeast(1)
    private val speakerVadParts = mutableListOf<FloatArray>()
    private val fallbackParts = mutableListOf<FloatArray>()
    private var speakerVadSampleCount = 0
    private var fallbackSampleCount = 0

    fun observe(samples: FloatArray, captureSpeakerVad: Boolean, captureFallback: Boolean) {
        if (captureSpeakerVad) {
            val retained = minOf(samples.size, maximumSamples - speakerVadSampleCount)
            if (retained > 0) {
                speakerVadParts += samples.copyOf(retained)
                speakerVadSampleCount += retained
            }
        }
        if (captureFallback) {
            val retained = minOf(samples.size, maximumSamples - fallbackSampleCount)
            if (retained > 0) {
                fallbackParts += samples.copyOf(retained)
                fallbackSampleCount += retained
            }
        }
    }

    fun speakerVadLength(): Int = speakerVadSampleCount

    fun speakerVadTail(sampleCount: Int): FloatArray {
        val retained = sampleCount.coerceIn(0, speakerVadSampleCount)
        val output = FloatArray(retained)
        var destinationEnd = retained
        for (index in speakerVadParts.indices.reversed()) {
            if (destinationEnd == 0) break
            val part = speakerVadParts[index]
            val copied = minOf(part.size, destinationEnd)
            System.arraycopy(
                part,
                part.size - copied,
                output,
                destinationEnd - copied,
                copied,
            )
            destinationEnd -= copied
        }
        return output
    }

    fun fallbackSamples(): FloatArray = concatenate(fallbackParts, fallbackSampleCount)

    fun clearNativeSegment() {
        speakerVadParts.clear()
        speakerVadSampleCount = 0
    }

    fun clearPublicUtterance() {
        fallbackParts.clear()
        fallbackSampleCount = 0
    }

    fun clearAll() {
        clearNativeSegment()
        clearPublicUtterance()
    }

    private fun concatenate(parts: List<FloatArray>, total: Int): FloatArray {
        val output = FloatArray(total)
        var offset = 0
        for (part in parts) {
            System.arraycopy(part, 0, output, offset, part.size)
            offset += part.size
        }
        return output
    }
}
